modules/ppcp-vaulting/src/PaymentTokenChecker.php

```diff
@@ -0,0 +1,225 @@
+
+	/**
+	 * The payments endpoint.
+	 *
+	 * @var PaymentsEndpoint
+	 */
+	protected $payments_endpoint;
+
+	/**
+	 * The logger.
+	 *
+	 * @var LoggerInterface
+	 */
+	protected $logger;
+
+	/**
+	 * PaymentTokenChecker constructor.
+	 *
+	 * @param PaymentTokenRepository      $payment_token_repository The payment token repository.
+	 * @param Settings                    $settings The settings.
+	 * @param AuthorizedPaymentsProcessor $authorized_payments_processor The authorized payments processor.
+	 * @param OrderEndpoint               $order_endpoint The order endpoint.
+	 * @param PaymentsEndpoint            $payments_endpoint The payments endpoint.
+	 * @param LoggerInterface             $logger The logger.
+	 */
+	public function __construct(
+		PaymentTokenRepository $payment_token_repository,
+		Settings $settings,
+		AuthorizedPaymentsProcessor $authorized_payments_processor,
+		OrderEndpoint $order_endpoint,
+		PaymentsEndpoint $payments_endpoint,
+		LoggerInterface $logger
+	) {
+		$this->payment_token_repository      = $payment_token_repository;
+		$this->settings                      = $settings;
+		$this->authorized_payments_processor = $authorized_payments_processor;
+		$this->order_endpoint                = $order_endpoint;
+		$this->payments_endpoint             = $payments_endpoint;
+		$this->logger                        = $logger;
+	}
+
+	/**
+	 * Check if payment token exist and updates order accordingly.
+	 *
+	 * @param int $order_id The order ID.
+	 * @param int $customer_id The customer ID.
+	 * @return void
+	 */
+	public function check_and_update( int $order_id, int $customer_id ):void {
+		$wc_order = wc_get_order( $order_id );
+		if ( ! is_a( $wc_order, WC_Order::class ) ) {
+			return;
+		}
+
+		if ( $wc_order->get_status() === 'processing' ) {
+			return;
+		}
+
+		$tokens = $this->payment_token_repository->all_for_user_id( $customer_id );
+		if ( $tokens ) {
+			try {
+				$this->capture_authorized_payment( $wc_order );
+			} catch ( NotFoundException $exception ) {
+				$this->logger->warning( "It was not possible to capture the payment for order: #{$order_id}" );
+			}
+
+			return;
+		}
+
+		$this->logger->error( "Payment for subscription parent order #{$order_id} was not saved on PayPal." );
+
+		try {
+			$order = $this->get_order( $wc_order );
+			$this->void_authorizations( $order );
+		} catch ( RuntimeException $exception ) {
+			$this->logger->warning( $exception->getMessage() );
+		}
+
+		$this->update_failed_status( $wc_order );
+	}
+
+	/**
+	 * Captures authorized payments for the given WC order.
+	 *
+	 * @param WC_Order $wc_order The WC order.
+	 * @throws NotFoundException When there is a problem capturing the payment.
+	 */
+	private function capture_authorized_payment( WC_Order $wc_order ): void {
+		if ( $this->settings->has( 'intent' ) && strtoupper( (string) $this->settings->get( 'intent' ) ) === 'CAPTURE' ) {
+			$this->authorized_payments_processor->capture_authorized_payment( $wc_order );
```

Comment:
Looks like `capture_authorized_payment` returns `bool`. Maybe should either return it too or check and throw.